### PR TITLE
copr: impl insert&insert_binary function

### PR DESCRIPTION
**Pull request by lerencao**
## What have you changed? (mandatory)

This PR implements `insert` and `insert_binary` scalar functions. 

Please **NOTE** that:

Different from the impl in TiDB, `maxAllowedPacket` is not handled in this PR.

## What are the type of the changes? (mandatory)
 New feature (change which adds functionality)
- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

- Unit Test
- make dev

## Does this PR affect documentation (docs) update? (mandatory)

Nope

## Does this PR affect tidb-ansible update? (mandatory)

Nope

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 1]
```diff
@@ -3427,4 +3503,78 @@ mod tests {
...
+            ("你好，", 2, 0, "世界", "你世界好，"),
+            ("你好，", 2, 1, "世界", "你世界，"),
+            ("你好，", 2, 2, "世界", "你世界"),
+            ("你好，", 2, 3, "世界", "你世界"),
+            ("你好，", 3, -1, "世界", "你好世界"),
+            ("你好，", 3, 0, "世界", "你好世界，"),
+            ("你好，", 3, 1, "世界", "你好世界"),
+            ("你好，", 3, 2, "世界", "你好世界"),
+        ];
+        for (s, pos, len, new_str, expected) in cases {
```
Comment (rleungx) [inline src/coprocessor/dag/expr/builtin_string.rs]: Can you also test if one of the arguments is `NULL`?

Comment (lerencao) [reply src/coprocessor/dag/expr/builtin_string.rs]: @rleungx Updated.

**Comment by rleungx** [comment]
@AndreMouche PTAL.

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 2]
```diff
@@ -944,6 +945,78 @@ impl ScalarFunc {
...
+            self.children[1].is_unsigned(),
+        );
+        let (mut len, len_positive) = i64_to_usize(
+            try_opt!(self.children[2].eval_int(ctx, row)),
+            self.children[2].is_unsigned(),
+        );
+        let new_str: Cow<[u8]> = try_opt!(self.children[3].eval_string(ctx, row));
+
+        // 1 <= pos <= s.len()
+        if !(pos_positive && (1..=s.len()).contains(&pos)) {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_string.rs]: It is too confusing.. why not use `1<s.len()|| s.len()>pos` directly?

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 3]
```diff
@@ -944,6 +945,78 @@ impl ScalarFunc {
...
+            return Ok(Some(s));
+        }
+        // 0 <= len <= s.len() - (pos - 1)
+        if !(len_positive && (0..=s.len() - pos + 1).contains(&len)) {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_string.rs]: ditto, why not use `len_positive>s.len()-pos + 1 || len_positive < 0` directly?

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 4]
```diff
@@ -944,6 +945,78 @@ impl ScalarFunc {
...
+
+    // See https://dev.mysql.com/doc/refman/5.6/en/string-functions.html#function_insert
+    #[inline]
+    pub fn insert_binary<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s: Cow<[u8]> = try_opt!(self.children[0].eval_string(ctx, row));
+        let (pos, pos_positive) = i64_to_usize(
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_string.rs]: Why not use `eval_int` directly?

**Comment by lerencao** [comment]
@AndreMouche 
Updated. PTAL

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 5]
```diff
@@ -944,6 +945,71 @@ impl ScalarFunc {
...
+    // See https://dev.mysql.com/doc/refman/5.6/en/string-functions.html#function_insert
+    #[inline]
+    pub fn insert_binary<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s: Cow<[u8]> = try_opt!(self.children[0].eval_string(ctx, row));
+        let pos = try_opt!(self.children[1].eval_int(ctx, row));
+        let mut len = try_opt!(self.children[2].eval_int(ctx, row));
```
[the remark this answers is not in the document]
Comment (lerencao) [reply src/coprocessor/dag/expr/builtin_string.rs]: In TiDB implementation, when constant evals to int, `aaa` is casted to `0`, which results to the output you gave in mysql/tidb .
While in TiKV coprocessor, when eval to int, bytes value is not handled, and error returns, and the behaviors is different.

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 6]
```diff
@@ -944,6 +945,71 @@ impl ScalarFunc {
...
+        let new_str: Cow<[u8]> = try_opt!(self.children[3].eval_string(ctx, row));
+
+        let s_len = s.len() as i64;
+        if pos < 1 || pos > s_len {
+            return Ok(Some(s));
+        }
+        if len < 0 || len > s_len - pos + 1 {
```
Comment (rleungx) [inline src/coprocessor/dag/expr/builtin_string.rs]: How about using a variable for  `s_len - pos + 1`?

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 7]
```diff
@@ -944,6 +945,71 @@ impl ScalarFunc {
...
+        let new_str: Cow<str> = try_opt!(self.children[3].eval_string_and_decode(ctx, row));
+
+        let char_len = s.chars().count() as i64;
+        if pos < 1 || pos > char_len {
+            return match s {
+                Cow::Borrowed(v) => Ok(Some(Cow::Borrowed(v.as_bytes()))),
+                Cow::Owned(v) => Ok(Some(Cow::Owned(v.into_bytes()))),
+            };
+        }
+        if len < 0 || len > char_len - pos + 1 {
```
Comment (rleungx) [inline src/coprocessor/dag/expr/builtin_string.rs]: ditto

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (AndreMouche) [reply src/coprocessor/dag/expr/builtin_string.rs]: Here comes TiDB's implementation
```
pos, isNull, err := b.args[1].EvalInt(b.ctx, row)
	if isNull || err != nil {
		return "", true, err
	}
	if pos < 1 || pos > runeLength {
		return str, false, nil
	}

	length, isNull, err := b.args[2].EvalInt(b.ctx, row)
	if isNull || err != nil {
		return "", true, err
	}
  ...
```
Seems we needn't consider the third argument when meeting the pos outside of the range.

[the remark this answers is not in the document]
Comment (lerencao) [reply src/coprocessor/dag/expr/builtin_string.rs]: Well, problem still remains, like the following outputs from TiDB.

```
mysql> select insert("xxxx","1",0,"bbb");
+----------------------------+
| insert("xxxx","1",0,"bbb") |
+----------------------------+
| bbbxxxx                    |
+----------------------------+
1 row in set (0.00 sec)

mysql> select insert("xxxx","1b",0,"bbb");
+-----------------------------+
| insert("xxxx","1b",0,"bbb") |
+-----------------------------+
| bbbxxxx                     |
+-----------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select insert("xxxx","b1",0,"bbb");
+-----------------------------+
| insert("xxxx","b1",0,"bbb") |
+-----------------------------+
| xxxx                        |
+-----------------------------+
1 row in set, 1 warning (0.00 sec)
```

But, I think I got your point. How TiKV eval string to int is another story which may be handled in there. 
The code is updated to follow the logic in TiDB.

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 6 above)
Comment (lerencao) [reply src/coprocessor/dag/expr/builtin_string.rs]: updated.

**Comment by rleungx** [comment]
@AndreMouche PTAL.

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 8]
```diff
@@ -933,6 +934,69 @@ impl ScalarFunc {
...
+        let remaining_len = s_len - pos + 1;
+        if len < 0 || len > remaining_len {
+            len = remaining_len
+        }
+        let new_str: Cow<'_, [u8]> = try_opt!(self.children[3].eval_string(ctx, row));
+
+        // now, pos > 0 and len >=0.
+        let mut t = s.to_vec();
+        drop(t.splice(
+            (pos - 1) as usize..(pos - 1 + len) as usize,
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_string.rs]: Do we need to consider `AllowedPacketOverflowed` as `tidb` did? @zz-jason

**Comment by zhouqiang-cl** [comment]
@AndreMouche @zz-jason How should this pr progress

**Comment by AndreMouche** [comment]
Please fix the conflicts, thank you

**Comment by AndreMouche** [comment]
We really value your contribution, we are closing this PR due to inactivity (last activity: 2019-04-09). We'd like to work with you more, if you are interested on continuing work on this PR please comment and we will re-open. If you'd like to work on something else, get in touch.
